src/plugins/lua/bindings/lua_ImGuiCustom.cpp

```diff
@@ -99,6 +102,39 @@ void RegisterBindings_ImGuiCustom(sol::table& ImGui)
 			[](mq::imgui::ConsoleWidget* pThis, const ImVec4& col, std::string_view text) { pThis->AppendText(text, MQColor(col), false); }
 		)
 	);
-}
 
+	// Widgets: ZepInputBox
+	ImGui.new_usertype<mq::imgui::ImGuiZepEditor>(
+		"TextEditor", sol::factories(&mq::imgui::ImGuiZepEditor::Create),
+		"Render", sol::overload(
+			[](mq::imgui::ImGuiZepEditor* pThis) { pThis->Render(); },
+			[](mq::imgui::ImGuiZepEditor* pThis, ImVec2 displaySize) { pThis->Render(displaySize); }
+		),
+		"Clear", &mq::imgui::ImGuiZepEditor::Clear,
+		"SetSyntax", [](mq::imgui::ImGuiZepEditor* pThis, const char* syntaxName) { pThis->SetSyntaxProvider(syntaxName); },
+		"text", sol::property(
+			[](mq::imgui::ImGuiZepEditor* pThis) -> std::string { size_t bufferSize = pThis->GetTextLength(); std::string text; text.reserve(bufferSize); pThis->GetText(text); return text; }/* read only */),
+		"LoadContents", [](mq::imgui::ImGuiZepEditor* pThis, std::string_view text) { pThis->SetText(text); },
+		"fontSize", sol::property(&mq::imgui::ImGuiZepEditor::GetFontSize, &mq::imgui::ImGuiZepEditor::SetFontSize),
+		"windowFlags", sol::property(&mq::imgui::ImGuiZepEditor::GetWindowFlags, &mq::imgui::ImGuiZepEditor::SetWindowFlags)
+	);
```

Comment:
I see that SetText is gone now, so I'm very curious what an example of use is for this widget since `text` is readonly (I think? I thought there was a sol::readonly decorator needed for that). We likely need to see an example to make sure the API is sensible before we can really comment on the actual bindings.

Reply:
I mean, i was fine with SetText/GetText but you told him to change it.

readonly and a property with only a setter are the same.

An example use case is pulling the text from the buffer to execute it as a lua script.

Reply:
That's not what I said, but we'll never know because of the force pushes erasing the code I commented on. I said that the implementation was leaking and I was wondering if there was a more user friendly way to implement this. I still think there's a very good way to not have to 1) detect text changes, 2) set it, and then 3) render it. #1 and #2 here are going to get messed up in userspace a lot.

Reply:
Just re-read the comment, I missed some bits because the formatting was messed up. I'm talking about this specifically:
` TextEditor:LoadContents(text) -- set the contents of the edit box, this is expensive and should be used sparingly`

I don't think it's wise to just throw this out to the user and hope it gets used correctly. What if you get a "buffer" that is actually a userdata with the string operations specified so that it naturally detects when the content changes each frame and then calls all of this behind the scenes?

Reply:
first of all, it is not an edit box. it is a text editor. It is supposed to be simple.

The buffer is not meant to be modified outside of the editor. Doing so is dangerous and requires a bunch of additional work (like managing glyph iterators and such)

Reply:
I think this is the bit that is completely unclear. When a user types in the text editor, what happens? The internal buffer gets modified? All of this API currently makes it seem like you need to manually call SetText (LoadContents -- a name change didn't change the functionality) every time the user changes text in the editor. Is that not how it works? If not, then this API needs to be made a lot clearer, because that's contrary to how most of ImGui works and users _will_ get confused

Reply:
When the user types into the editor, it changes the contents of the editor. You can serialize that out as a string (.text) or load in new contents to be parsed into the internal buffer and display in the editor

Reply:
We don't currently have a text editor to expose to lua, zep does, and it can't be used directly in lua because zep isn't available to plugins, but that's the other point -- you can't do this work _without_ creating a new imgui widget.

and you are very right that no one has written documentation, I think that we should have had at least an example of what this is supposed to look like first, which is what I said in the first place.

Reply:
> We don't currently have a text editor to expose to lua, zep does, and it can't be used directly in lua because zep isn't available to plugins, but that's the other point -- you can't do this work _without_ creating a new imgui widget.
> 
> and you are very right that no one has written documentation, I think that we should have had at least an example of what this is supposed to look like first, which is what I said in the first place.

[texteditor.lua.txt](https://github.com/user-attachments/files/17353207/texteditor.lua.txt)

Reply:
@grimmier378 turn that into a gist. pasting a text file around sucks.

Reply:
[<script src="https://gist.github.com/grimmier378/be326fd3f67d009cb2f7c93128c45369.js"></script>](https://gist.github.com/grimmier378/be326fd3f67d009cb2f7c93128c45369)